src/bootloaders/shim-systemd.c

```diff
@@ -348,16 +346,24 @@ static bool shim_systemd_init(const BootManager *manager)
         systemd_src = string_printf("%s/%s", prefix, SYSTEMD_SRC);
 
         boot_root = boot_manager_get_boot_dir((BootManager *)manager);
-        len = strlen(boot_root);
-        /* SHIM_DST and SYSTEMD_DST are defined with leading '/', take extra
-         * care to produce clean paths. */
-        if (len > 0 && boot_root[len - 1] == '/') {
-                boot_root[len - 1] = '\0';
-        }
-        shim_dst_host = string_printf("%s%s", boot_root, SHIM_DST);
-        systemd_dst_host = string_printf("%s%s", boot_root, SYSTEMD_DST);
-
-        shim_dst_esp = SHIM_DST;
+        shim_dst_host = nc_build_case_correct_path(boot_root,
+                                                   ESP_EFI,
+                                                   KERNEL_NAMESPACE,
+                                                   SHIM_DST,
+                                                   NULL);
+        systemd_dst_host = nc_build_case_correct_path(boot_root,
+                                                      ESP_EFI,
+                                                      KERNEL_NAMESPACE,
+                                                      SYSTEMD_DST,
+                                                      NULL);
+
+        shim_dst_esp = strdup(shim_dst_host + strlen(boot_root));
+
+        efi_fallback_dst_host =
+            nc_build_case_correct_path(boot_root, ESP_EFI, ESP_BOOT, EFI_FALLBACK, NULL);
+        kernel_dst_host =
+            nc_build_case_correct_path(boot_root, ESP_EFI, KERNEL_NAMESPACE, KERNEL_DST_DIR, NULL);
+        kernel_dst_esp = strdup(kernel_dst_host + strlen(boot_root));
```

Comment:
...?

Reply:
oh, yeah, this is very ugly. it's basically to get absolute path on ESP.

i am thinking that the right solution to avoid this would be to return full mounted path and then use it in `kernel.c`. right now, it [constructs its own path](https://github.com/clearlinux/clr-boot-manager/blob/master/src/bootman/kernel.c#L613) while it could use full path if `get_kernel_destination()` returned full path. we will then not need to do this and `kernel.c` constructing its own path. what do you think?